dwave/optimization/src/nodes/creation.cpp

```diff
@@ -0,0 +1,309 @@
+}
+
+bool ARangeNode::integral() const { return true; }
+
+void ARangeNode::initialize_state(State& state) const {
+    emplace_data_ptr<ArrayNodeStateData>(state, arange(state, start_, stop_, step_));
+}
+
+std::pair<double, double> ARangeNode::minmax(
+        optional_cache_type<std::pair<double, double>> cache) const {
+    return memoize(cache, [&]() {
+        auto visitor = get_minmax();
+        const auto [start_low, start_high] = std::visit(visitor, start_);
+        const auto [stop_low, stop_high] = std::visit(visitor, stop_);
+        const auto [step_low, step_high] = std::visit(visitor, step_);
+
+        // checked on construction.
+        assert(!(step_low <= 0 && step_high >= 0));
+
+        // The direction that we're stepping determines the min/max
+        if (step_low > 0) {
+            const double low = start_low;
+            const double high = stop_high;
+            if (low > high) return std::pair<double, double>(0, 0);
+            return std::make_pair(low, high);
+        }
+        if (step_low < 0) {
+            const double low = stop_low;
+            const double high = start_high;
+            if (low > high) return std::pair<double, double>(0, 0);
+            return std::make_pair(low, high);
+        }
+
+        assert(false && "zero step not allowed");
+        unreachable();
+    });
+}
+
+void ARangeNode::propagate(State& state) const {
+    auto visitor = get_diff(state);
+    const auto [start_old, start_new] = std::visit(visitor, start_);
+    const auto [stop_old, stop_new] = std::visit(visitor, stop_);
+    const auto [step_old, step_new] = std::visit(visitor, step_);
+
+    ArrayNodeStateData* ptr = data_ptr<ArrayNodeStateData>(state);
+
+    // If the start or the step has changed, we need to change everything
+    // Alternatively if there is currently nothing in the buffer at all, we might
+    // as well just assign
+    if (start_old != start_new || step_old != step_new || ptr->size() == 0) {
+        ptr->assign(arange(state, start_, stop_, step_));
+        if (ptr->diff().size()) Node::propagate(state);
+        return;
+    }
+
+    // if the stop has also not changed, then nothing to do
+    if (stop_old == stop_new) return;
+
+    // only the stop has changed, and it has actually changed
+    const ssize_t step = step_old;
+    assert(step == step_new);
+
+    const auto& buffer = ptr->buffer;  // readonly for our purposes
+    assert(buffer.size() >= 1);
+
+    if (step > 0) {
+        if (stop_old < stop_new) {
+            // we grew
+            for (const double& val : arange(buffer.back() + step, stop_new, step)) {
+                ptr->emplace_back(val);
+            }
+        } else {
+            // we shrank
+            while (buffer.size() && buffer.back() >= stop_new) {
+                ptr->pop_back();
+            }
+        }
+    } else if (step < 0) {
+        if (stop_old < stop_new) {
+            // we shank
```

Comment:
Typo

Reply:
In hindsight this is very funny

Reply:
probably good arbitrary reacts aren't supported in github, but :hocho: